src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentRecorder.cs

```diff
@@ -0,0 +1,184 @@
+    /// </summary>
+    public sealed class InstrumentRecorder<T> : IDisposable where T : struct
+    {
+        private bool _isObservableInstrument;
+
+        private readonly MeterListener _meterListener;
+        private readonly List<Measurement<T>> _measurements;
+
+        /// <summary>
+        /// Initialize a new instance <see cref="InstrumentRecorder{T}" /> to record the measurements published by <see cref="Instrument" />.
+        /// </summary>
+        /// <param name="instrument">The <see cref="Instrument" /> to record measurements from.</param>
+        public InstrumentRecorder(Instrument instrument)
+        {
+            if (instrument is null)
+            {
+                throw new ArgumentNullException(nameof(instrument));
+            }
+
+            if (instrument is not Instrument<T> and not ObservableInstrument<T>)
+            {
+                throw new InvalidOperationException(SR.InvalidInstrumentType);
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (inst, listener) =>
+            {
+                if (object.ReferenceEquals(inst, instrument))
+                {
+                    if (instrument is ObservableInstrument<T>)
+                    {
+                        _isObservableInstrument = true;
+                    }
+
+                    listener.EnableMeasurementEvents(inst, state: null);
+
+                    Instrument = instrument;
+                }
+            };
+
+            _meterListener.SetMeasurementEventCallback<T>(OnMeasurementRecorded);
+            _meterListener.Start();
+        }
+
+        /// <summary>
+        /// Initialize a new instance <see cref="InstrumentRecorder{T}" /> to record the measurement published by an <see cref="Instrument" /> with the name <paramref name="instrumentName" />
+        /// and a meter having the name <paramref name="meterName"/> and scope <paramref name="scopeFilter"/>
+        /// </summary>
+        /// <param name="scopeFilter">The name of the meter.</param>
+        /// <param name="meterName">The name of the meter.</param>
+        /// <param name="instrumentName">The name of the instrument.</param>
+        public InstrumentRecorder(object? scopeFilter, string meterName, string instrumentName)
+        {
+            if (meterName is null)
+            {
+                throw new ArgumentNullException(nameof(meterName));
+            }
+
+            if (instrumentName is null)
+            {
+                throw new ArgumentNullException(nameof(instrumentName));
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (instrument, listener) =>
+            {
```

Comment:
Is it possible this callback might be invoked multiple times for the same meterName / instrumentName?  I'm wondering if you need to guard against that, or if not, assert it's not the case, since it looks like things might go a bit haywire if that were to ever happen.